File: src/main/java/com/freya02/botcommands/prefixed/impl/BaseCommandEventImpl.java
package com.freya02.botcommands.prefixed.impl;

import com.freya02.botcommands.BContext;
import com.freya02.botcommands.internal.BContextImpl;
import com.freya02.botcommands.internal.Logging;
import com.freya02.botcommands.prefixed.BaseCommandEvent;
import com.freya02.botcommands.utils.EmojiUtils;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.MessageChannel;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.events.message.guild.GuildMessageReceivedEvent;
import net.dv8tion.jda.api.requests.RestAction;
import net.dv8tion.jda.api.utils.AttachmentOption;
import org.slf4j.Logger;

import javax.annotation.CheckReturnValue;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.InputStream;
import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

public class BaseCommandEventImpl extends BaseCommandEvent {
	public static final String SUCCESS = EmojiUtils.resolveEmojis(":white_check_mark:");
	public static final String ERROR = EmojiUtils.resolveEmojis(":x:");
	private static final Logger LOGGER = Logging.getLogger();

	protected final BContext context;
	protected final String argumentsStr;

	public BaseCommandEventImpl(BContext context, GuildMessageReceivedEvent event, String arguments) {
		super(event.getJDA(), event.getResponseNumber(), event.getMessage());

		this.context = context;
		this.argumentsStr = arguments;
	}

	@Override
	public BContext getContext() {
		return context;
	}

	@Override
	public List<String> getArgumentsStrList() {
		if (!argumentsStr.isBlank()) {
			return Arrays.asList(argumentsStr.split(" "));
		} else {
			return List.of();
		}
	}

	@Override
	public String getArgumentsStr() {
		return argumentsStr;
	}

	@Override
	public void reportError(String message, Throwable e) {
		channel.sendMessage(message).queue(null, t -> LOGGER.error("Could not send message to channel : {}", message));

		((BContextImpl) context).dispatchException(message, e);
	}

	@Override
	public Consumer<? super Throwable> failureReporter(String message) {
		return t -> reportError(message, t);
	}

	@Override
	public String getAuthorBestName() {
		return getMember().getEffectiveName();
	}

	@Override
	@Nonnull
	public EmbedBuilder getDefaultEmbed() {
		return context.getDefaultEmbedSupplier().get();
	}

	@Override
	@Nullable
	public InputStream getDefaultIconStream() {
		return context.getDefaultFooterIconSupplier().get();
	}

	@Override
	public RestAction<Message> sendWithEmbedFooterIcon(MessageEmbed embed, Consumer<? super Throwable> onException) {
		return sendWithEmbedFooterIcon(channel, embed, onException);
	}

	@Override
	@CheckReturnValue
	public RestAction<Message> sendWithEmbedFooterIcon(MessageChannel channel, MessageEmbed embed, Consumer<? super Throwable> onException) {
		return sendWithEmbedFooterIcon(channel, getDefaultIconStream(), embed, onException);
	}

	@Override
	@CheckReturnValue
	public RestAction<Message> sendWithEmbedFooterIcon(MessageChannel channel, InputStream iconStream, MessageEmbed embed, Consumer<? super Throwable> onException) {
		if (iconStream != null) {
			return channel.sendTyping().flatMap(v -> channel.sendFile(iconStream, "icon.jpg").setEmbeds(embed));
		} else {
			return channel.sendTyping().flatMap(v -> channel.sendMessageEmbeds(embed));
		}
	}

	@Override
	@CheckReturnValue
	public RestAction<Void> reactSuccess() {
		return channel.addReactionById(messageId, SUCCESS);
	}

	@Override
	@CheckReturnValue
	public RestAction<Void> reactError() {
		return channel.addReactionById(messageId, ERROR);
	}

	@Nonnull
	@Override
	public RestAction<Message> respond(@Nonnull CharSequence text) {
		return channel.sendMessage(text);
	}

	@Nonnull
	@Override
	public RestAction<Message> respondFormat(@Nonnull String format, @Nonnull Object... args) {
		return channel.sendMessageFormat(format, args);
	}

	@Nonnull
	@Override
	public RestAction<Message> respond(@Nonnull MessageEmbed embed, @Nonnull MessageEmbed... other) {
		return channel.sendMessageEmbeds(embed, other);
	}

	@Nonnull
	@Override
	public RestAction<Message> respondFile(@Nonnull InputStream data, @Nonnull String fileName, @Nonnull AttachmentOption... options) {
		return channel.sendFile(data, fileName, options);
	}

	@Nonnull
	@Override
	public RestAction<Message> respondFile(@Nonnull byte[] data, @Nonnull String fileName, @Nonnull AttachmentOption... options) {
		return channel.sendFile(data, fileName, options);
	}

	@Override
	@CheckReturnValue
	@Nonnull
	public RestAction<Message> reply(@Nonnull CharSequence text) {
		return getMessage().reply(text);
	}

	@Override
	@CheckReturnValue
	@Nonnull
	public RestAction<Message> replyFormat(@Nonnull String format, @Nonnull Object... args) {
		return getMessage().replyFormat(format, args);
	}

	@Override
	@CheckReturnValue
	@Nonnull
	public RestAction<Message> reply(@Nonnull MessageEmbed embed, @Nonnull MessageEmbed... other) {
		return getMessage().replyEmbeds(embed, other);
	}

	@Override
	@CheckReturnValue
	@Nonnull
	public RestAction<Message> replyFile(@Nonnull InputStream data, @Nonnull String fileName, @Nonnull AttachmentOption... options) {
		return channel.sendTyping().flatMap(v -> getMessage().reply(data, fileName, options));
	}

	@Override
	@CheckReturnValue
	@Nonnull
	public RestAction<Message> replyFile(byte[] data, @Nonnull String fileName, @Nonnull AttachmentOption... options) {
		return channel.sendTyping().flatMap(v -> getMessage().reply(data, fileName, options));
	}

	@Nonnull
	@Override
	public RestAction<Message> indicateError(@Nonnull CharSequence text) {
		if (getGuild().getSelfMember().hasPermission(channel, Permission.MESSAGE_ADD_REACTION)) {
			return reactError().flatMap(v -> channel.sendMessage(text));
		} else {
			return channel.sendMessage(text);
		}
	}

	@Nonnull
	@Override
	public RestAction<Message> indicateErrorFormat(@Nonnull String format, @Nonnull Object... args) {
		if (getGuild().getSelfMember().hasPermission(channel, Permission.MESSAGE_ADD_REACTION)) {
			return reactError().flatMap(v -> channel.sendMessageFormat(format, args));
		} else {
			return channel.sendMessageFormat(format, args);
		}
	}

	@Nonnull
	@Override
	public RestAction<Message> indicateError(@Nonnull MessageEmbed embed, @Nonnull MessageEmbed... other) {
		if (getGuild().getSelfMember().hasPermission(channel, Permission.MESSAGE_ADD_REACTION)) {
			return reactError().flatMap(v -> channel.sendMessageEmbeds(embed, other));
		} else {
			return channel.sendMessageEmbeds(embed, other);
		}
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/commands/prefixed/BaseCommandEventImpl.kt
package com.freya02.botcommands.internal.commands.prefixed

import com.freya02.botcommands.api.commands.prefixed.BaseCommandEvent
import com.freya02.botcommands.api.core.utils.logger
import com.freya02.botcommands.api.utils.EmojiUtils
import com.freya02.botcommands.internal.BContextImpl
import mu.KotlinLogging
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import net.dv8tion.jda.api.requests.RestAction
import net.dv8tion.jda.api.utils.FileUpload
import java.io.InputStream
import java.util.function.Consumer
import javax.annotation.CheckReturnValue

private val logger = KotlinLogging.logger<BaseCommandEvent>()

internal open class BaseCommandEventImpl(
    context: BContextImpl,
    event: MessageReceivedEvent,
    arguments: String
) : BaseCommandEvent(context, event.jda, event.responseNumber, event.message) {
    private val argumentsStr: String = arguments

    override fun getArgumentsStrList(): List<String> = when {
        argumentsStr.isNotBlank() -> argumentsStr.split(' ').dropLastWhile { it.isEmpty() }
        else -> listOf()
    }

    override fun getArgumentsStr(): String = argumentsStr

    override fun reportError(message: String, e: Throwable) {
        channel.sendMessage(message).queue(null) { t: Throwable? -> logger.error("Could not send message to channel : {}", message, t) }
        context.dispatchException(message, e)
    }

    override fun failureReporter(message: String): Consumer<in Throwable> {
        return Consumer { t: Throwable -> reportError(message, t) }
    }

    override fun getAuthorBestName(): String {
        return member.effectiveName
    }

    override fun getDefaultEmbed(): EmbedBuilder {
        return context.defaultEmbedSupplier.get()
    }

    override fun getDefaultIconStream(): InputStream? = context.defaultFooterIconSupplier.get()

    override fun sendWithEmbedFooterIcon(embed: MessageEmbed, onException: Consumer<in Throwable>): RestAction<Message> =
        sendWithEmbedFooterIcon(channel, embed, onException)

    @CheckReturnValue
    override fun sendWithEmbedFooterIcon(
        channel: MessageChannel,
        embed: MessageEmbed,
        onException: Consumer<in Throwable>
    ): RestAction<Message> = sendWithEmbedFooterIcon(channel, defaultIconStream, embed, onException)

    @CheckReturnValue
    override fun sendWithEmbedFooterIcon(
        channel: MessageChannel,
        iconStream: InputStream?,
        embed: MessageEmbed,
        onException: Consumer<in Throwable>
    ): RestAction<Message> = when {
        iconStream != null -> channel.sendTyping().flatMap { channel.sendFiles(FileUpload.fromData(iconStream, "icon.jpg")).setEmbeds(embed) }
        else -> channel.sendTyping().flatMap { channel.sendMessageEmbeds(embed) }
    }

    @CheckReturnValue
    override fun reactSuccess(): RestAction<Void> = channel.addReactionById(messageId, SUCCESS)

    @CheckReturnValue
    override fun reactError(): RestAction<Void> = channel.addReactionById(messageId, ERROR)

    override fun respond(text: CharSequence): RestAction<Message> = channel.sendMessage(text)

    override fun respondFormat(format: String, vararg args: Any): RestAction<Message> = channel.sendMessageFormat(format, *args)

    override fun respond(embed: MessageEmbed, vararg other: MessageEmbed): RestAction<Message> = channel.sendMessageEmbeds(embed, *other)

    override fun respondFile(vararg fileUploads: FileUpload): RestAction<Message> = channel.sendFiles(*fileUploads)

    @CheckReturnValue
    override fun reply(text: CharSequence): RestAction<Message> = message.reply(text)

    @CheckReturnValue
    override fun replyFormat(format: String, vararg args: Any): RestAction<Message> = message.replyFormat(format, *args)

    @CheckReturnValue
    override fun reply(embed: MessageEmbed, vararg other: MessageEmbed): RestAction<Message> = message.replyEmbeds(embed, *other)

    @CheckReturnValue
    override fun replyFile(vararg fileUploads: FileUpload): RestAction<Message> =
        channel.sendTyping().flatMap { message.replyFiles(*fileUploads) }

    override fun indicateError(text: CharSequence): RestAction<Message> = when {
        guild.selfMember.hasPermission(guildChannel, Permission.MESSAGE_ADD_REACTION) -> reactError().flatMap { channel.sendMessage(text) }
        else -> channel.sendMessage(text)
    }

    override fun indicateErrorFormat(format: String, vararg args: Any): RestAction<Message> = when {
        guild.selfMember.hasPermission(guildChannel, Permission.MESSAGE_ADD_REACTION) -> reactError().flatMap { channel.sendMessageFormat(format, *args) }
        else -> channel.sendMessageFormat(format, *args)
    }

    override fun indicateError(embed: MessageEmbed, vararg other: MessageEmbed): RestAction<Message> = when {
        guild.selfMember.hasPermission(guildChannel, Permission.MESSAGE_ADD_REACTION) -> reactError().flatMap { channel.sendMessageEmbeds(embed, *other) }
        else -> channel.sendMessageEmbeds(embed, *other)
    }

    companion object {
        @JvmField val SUCCESS = EmojiUtils.resolveJDAEmoji(":white_check_mark:")
        @JvmField val ERROR = EmojiUtils.resolveJDAEmoji(":x:")
    }
}